File: sdk/scala/src/main/scala/bsdkrun/CI.scala
package bsdkrun

import java.nio.file.{Files, Path}
import scala.collection.immutable.SortedMap

/** CI workflows defined in code instead of YAML.
  *
  * The builder produces exactly the file `bsdkrun ci` (and tangled's spindle)
  * consumes — [[CIWorkflow.yaml]] is that file, [[CIWorkflow.save]] commits it
  * to `.tangled/workflows/`, and [[CIWorkflow.run]] executes it in a microVM
  * without a file ever touching the repository:
  *
  * {{{
  * CI.workflow("test")
  *   .onPush("main")
  *   .deps("scala", "sbt")
  *   .env("SBT_OPTS", "-Xmx1g")
  *   .step("compile", "sbt compile")
  *   .step("test", "sbt test")
  *   .run()
  * }}}
  *
  * Code is the source of truth and YAML the wire format, in that order —
  * which is why `save` writes a generated-file header: a hand-edit there will
  * be overwritten by the next save.
  */
object CI:
  /** Start a CI workflow definition. */
  def workflow(name: String): CIWorkflow = CIWorkflow(name = name)

final case class CIWorkflow(
    name: String,
    engine: String = "nixery",
    when: Vector[(Vector[String], Vector[String])] = Vector.empty,
    // Sorted for deterministic output: the emitted YAML is committed and
    // diffed, so its ordering must not depend on insertion order.
    dependencies: SortedMap[String, Vector[String]] = SortedMap.empty,
    environment: SortedMap[String, String] = SortedMap.empty,
    steps: Vector[CIStep] = Vector.empty,
    cloneDepth: Option[Int] = None,
    cloneSkip: Boolean = false
):
  /** Override the engine (`nixery` by default). */
  def withEngine(e: String): CIWorkflow = copy(engine = e)

  /** Add a push trigger for the given branches. */
  def onPush(branches: String*): CIWorkflow =
    copy(when = when :+ (Vector("push"), branches.toVector))

  /** Add a pull_request trigger targeting the given branches. */
  def onPullRequest(branches: String*): CIWorkflow =
    copy(when = when :+ (Vector("pull_request"), branches.toVector))

  /** Add nixpkgs dependencies — the toolchain the steps run against. */
  def deps(packages: String*): CIWorkflow = depsFrom("nixpkgs", packages*)

  /** Add dependencies from a custom registry (a flake reference). */
  def depsFrom(registry: String, packages: String*): CIWorkflow =
    copy(dependencies = dependencies.updated(
      registry,
      dependencies.getOrElse(registry, Vector.empty) ++ packages
    ))

  /** Set a workflow-level environment variable. */
  def env(key: String, value: String): CIWorkflow =
    copy(environment = environment.updated(key, value))

  /** Append a step; steps run serially in one VM, from the workspace root. */
  def step(name: String, command: String): CIWorkflow =
    copy(steps = steps :+ CIStep(name, command))

  /** Append a step with step-scoped environment variables. */
  def step(name: String, command: String, env: Map[String, String]): CIWorkflow =
    copy(steps = steps :+ CIStep(name, command, SortedMap.from(env)))

  /** Set the clone depth (default 1). */
  def withCloneDepth(depth: Int): CIWorkflow = copy(cloneDepth = Some(depth))

  /** Skip the checkout entirely. */
  def skipClone: CIWorkflow = copy(cloneSkip = true)

  /** The workflow file name [[save]] writes: `<name>.yml`. */
  def fileName: String =
    if name.endsWith(".yml") || name.endsWith(".yaml") then name
    else s"$name.yml"

  /** Render the workflow file.
    *
    * Scalars are emitted as JSON strings — valid YAML by construction — and
    * commands as literal blocks when safe, so the SDK needs no YAML
    * dependency.
    */
  def yaml: String =
    val sections = Vector.newBuilder[String]

    if when.nonEmpty then
      val lines = when.flatMap { case (events, branches) =>
        val head = s"  - event: [${events.map(q).mkString(", ")}]"
        branches.toList match
          case Nil        => Vector(head)
          case one :: Nil => Vector(head, s"    branch: ${q(one)}")
          case many => Vector(head, s"    branch: [${many.map(q).mkString(", ")}]")
      }
      sections += ("when:" +: lines).mkString("\n")

    sections += s"engine: $engine"

    if dependencies.nonEmpty then
      val lines = dependencies.toVector.flatMap { case (reg, pkgs) =>
        s"  ${q(reg)}:" +: pkgs.map(p => s"    - ${q(p)}")
      }
      sections += ("dependencies:" +: lines).mkString("\n")

    if environment.nonEmpty then
      val lines = environment.toVector.map { case (k, v) => s"  $k: ${q(v)}" }
      sections += ("environment:" +: lines).mkString("\n")

    if cloneSkip || cloneDepth.nonEmpty then
      val lines = Vector("clone:") ++
        (if cloneSkip then Vector("  skip: true") else Vector.empty) ++
        cloneDepth.map(d => s"  depth: $d").toVector
      sections += lines.mkString("\n")

    val stepLines = steps.flatMap { s =>
      Vector(s"  - name: ${q(s.name)}") ++ commandLines(s.command) ++
        (if s.env.nonEmpty then
           "    environment:" +: s.env.toVector.map { case (k, v) =>
             s"      $k: ${q(v)}"
           }
         else Vector.empty)
    }
    sections += ("steps:" +: stepLines).mkString("\n")

    sections.result().mkString("\n\n") + "\n"

  /** A literal block when it round-trips byte-for-byte; a JSON string when it
    * cannot (trailing spaces, carriage returns) — never a silent alteration.
    */
  private def commandLines(command: String): Vector[String] =
    val blockSafe = command.nonEmpty &&
      !command.contains('\r') &&
      command.split("\n", -1).forall(l => l == l.replaceAll(" +$", ""))
    if !blockSafe then Vector(s"    command: ${q(command)}")
    else
      "    command: |" +: command
        .replaceAll("\n+$", "")
        .split("\n", -1)
        .toVector
        .map(l => s"      $l")

  /** A JSON string literal, which is a valid YAML scalar by construction. */
  private def q(s: String): String =
    val sb = StringBuilder("\"")
    s.foreach {
      case '"'           => sb.append("\\\"")
      case '\\'          => sb.append("\\\\")
      case '\n'          => sb.append("\\n")
      case '\r'          => sb.append("\\r")
      case '\t'          => sb.append("\\t")
      case c if c < 0x20 => sb.append(f"\\u${c.toInt}%04x")
      case c             => sb.append(c)
    }
    sb.append('"').toString

  /** Write into `<repo>/.tangled/workflows/` and return the path. */
  def save(repo: Path): Path =
    val dir = repo.resolve(".tangled").resolve("workflows")
    Files.createDirectories(dir)
    val path = dir.resolve(fileName)
    Files.writeString(
      path,
      "# Generated by the bsdkrun SDK — edit the code that save()d it instead.\n" + yaml
    )
    path

  /** Execute the workflow in a microVM, streaming output.
    *
    * The YAML never touches the repository — it goes to a temp file and
    * `bsdkrun ci run -f`. Returns the failing step's error, or unit.
    */
  def run(dir: Option[Path] = None): Either[BsdkrunError, Unit] =
    val tmp = Files.createTempDirectory("bsdkrun-ci-")
    val file = tmp.resolve(fileName)
    Files.writeString(file, yaml)
    val args = Seq("ci", "run", "-f", file.toString) ++
      dir.toSeq.flatMap(d => Seq("-w", d.toString))
    val result = Proc.spawn(args)
    Files.deleteIfExists(file)
    Files.deleteIfExists(tmp)
    result.flatMap {
      case 0 => Right(())
      case code =>
        Left(
          BsdkrunError.CommandFailed(
            command = s"bsdkrun ci run ($name)",
            exitCode = code,
            stdout = "",
            stderr = s"workflow $name failed"
          )
        )
    }

final case class CIStep(
    name: String,
    command: String,
    env: SortedMap[String, String] = SortedMap.empty
)

File: sdk/elixir/lib/bsdkrun/ci.ex
defmodule Bsdkrun.CI do
  @moduledoc """
  CI workflows defined in code instead of YAML.

  The builder produces exactly the file `bsdkrun ci` (and tangled's spindle)
  consumes — `yaml/1` is that file, `save/2` commits it to
  `.tangled/workflows/`, and `run/1` executes it in a microVM without a file
  ever touching the repository:

      Bsdkrun.CI.workflow("test")
      |> Bsdkrun.CI.on_push("main")
      |> Bsdkrun.CI.deps(["elixir", "erlang"])
      |> Bsdkrun.CI.env("MIX_ENV", "test")
      |> Bsdkrun.CI.step("deps", "mix deps.get")
      |> Bsdkrun.CI.step("test", "mix test")
      |> Bsdkrun.CI.run()

  Code is the source of truth and YAML the wire format, in that order — which
  is why `save/2` writes a generated-file header: a hand-edit there will be
  overwritten by the next save.
  """

  defstruct name: nil,
            engine: "nixery",
            when_: [],
            deps: %{},
            env: %{},
            steps: [],
            clone_depth: nil,
            clone_skip: false

  @type t :: %__MODULE__{}

  @doc "Start a CI workflow definition."
  @spec workflow(String.t()) :: t()
  def workflow(name), do: %__MODULE__{name: name}

  @doc "Override the engine (`nixery` by default)."
  @spec engine(t(), String.t()) :: t()
  def engine(%__MODULE__{} = wf, engine), do: %{wf | engine: engine}

  @doc "Add a push trigger for the given branch(es)."
  @spec on_push(t(), String.t() | [String.t()]) :: t()
  def on_push(%__MODULE__{} = wf, branches),
    do: %{wf | when_: wf.when_ ++ [{["push"], List.wrap(branches)}]}

  @doc "Add a pull_request trigger targeting the given branch(es)."
  @spec on_pull_request(t(), String.t() | [String.t()]) :: t()
  def on_pull_request(%__MODULE__{} = wf, branches),
    do: %{wf | when_: wf.when_ ++ [{["pull_request"], List.wrap(branches)}]}

  @doc "Add nixpkgs dependencies — the toolchain the steps run against."
  @spec deps(t(), [String.t()]) :: t()
  def deps(%__MODULE__{} = wf, packages),
    do: %{wf | deps: Map.update(wf.deps, "nixpkgs", packages, &(&1 ++ packages))}

  @doc "Add dependencies from a custom registry (a flake reference)."
  @spec deps_from(t(), String.t(), [String.t()]) :: t()
  def deps_from(%__MODULE__{} = wf, registry, packages),
    do: %{wf | deps: Map.update(wf.deps, registry, packages, &(&1 ++ packages))}

  @doc "Set a workflow-level environment variable."
  @spec env(t(), String.t(), String.t()) :: t()
  def env(%__MODULE__{} = wf, key, value), do: %{wf | env: Map.put(wf.env, key, value)}

  @doc "Append a step; steps run serially in one VM, from the workspace root."
  @spec step(t(), String.t(), String.t(), %{String.t() => String.t()}) :: t()
  def step(%__MODULE__{} = wf, name, command, step_env \\ %{}),
    do: %{wf | steps: wf.steps ++ [%{name: name, command: command, env: step_env}]}

  @doc "Set the clone depth (default 1)."
  @spec clone_depth(t(), pos_integer()) :: t()
  def clone_depth(%__MODULE__{} = wf, depth), do: %{wf | clone_depth: depth}

  @doc "Skip the checkout entirely."
  @spec skip_clone(t()) :: t()
  def skip_clone(%__MODULE__{} = wf), do: %{wf | clone_skip: true}

  @doc "The workflow file name `save/2` writes: `<name>.yml`."
  @spec file_name(t()) :: String.t()
  def file_name(%__MODULE__{name: name}) do
    if String.ends_with?(name, [".yml", ".yaml"]), do: name, else: name <> ".yml"
  end

  @doc """
  Render the workflow file.

  Scalars are emitted as JSON strings — valid YAML by construction — and
  commands as literal blocks when safe, so the SDK needs no YAML dependency.
  """
  @spec yaml(t()) :: String.t()
  def yaml(%__MODULE__{} = wf) do
    sections =
      [
        when_section(wf),
        "engine: #{wf.engine}",
        deps_section(wf),
        env_section(wf),
        clone_section(wf),
        steps_section(wf)
      ]
      |> Enum.reject(&is_nil/1)

    Enum.join(sections, "\n\n") <> "\n"
  end

  defp when_section(%{when_: []}), do: nil

  defp when_section(%{when_: constraints}) do
    lines =
      Enum.flat_map(constraints, fn {events, branches} ->
        head = "  - event: [#{Enum.map_join(events, ", ", &q/1)}]"

        case branches do
          [] -> [head]
          [one] -> [head, "    branch: #{q(one)}"]
          many -> [head, "    branch: [#{Enum.map_join(many, ", ", &q/1)}]"]
        end
      end)

    Enum.join(["when:" | lines], "\n")
  end

  defp deps_section(%{deps: deps}) when map_size(deps) == 0, do: nil

  defp deps_section(%{deps: deps}) do
    lines =
      deps
      |> Map.keys()
      |> Enum.sort()
      |> Enum.flat_map(fn reg ->
        ["  #{q(reg)}:" | Enum.map(deps[reg], &"    - #{q(&1)}")]
      end)

    Enum.join(["dependencies:" | lines], "\n")
  end

  defp env_section(%{env: env}) when map_size(env) == 0, do: nil

  defp env_section(%{env: env}) do
    lines = env |> Map.keys() |> Enum.sort() |> Enum.map(&"  #{&1}: #{q(env[&1])}")
    Enum.join(["environment:" | lines], "\n")
  end

  defp clone_section(%{clone_skip: false, clone_depth: nil}), do: nil

  defp clone_section(%{clone_skip: skip, clone_depth: depth}) do
    lines =
      ["clone:"] ++
        if(skip, do: ["  skip: true"], else: []) ++
        if(depth, do: ["  depth: #{depth}"], else: [])

    Enum.join(lines, "\n")
  end

  defp steps_section(%{steps: steps}) do
    lines =
      Enum.flat_map(steps, fn s ->
        ["  - name: #{q(s.name)}"] ++
          command_lines(s.command) ++ step_env_lines(s.env)
      end)

    Enum.join(["steps:" | lines], "\n")
  end

  # Literal blocks read well in a committed file, but cannot carry trailing
  # spaces or carriage returns byte-for-byte; fall back to a JSON string
  # rather than silently altering the command.
  defp command_lines(command) do
    block_safe? =
      command != "" and
        not String.contains?(command, "\r") and
        command |> String.split("\n") |> Enum.all?(&(&1 == String.trim_trailing(&1, " ")))

    if block_safe? do
      body =
        command
        |> String.trim_trailing("\n")
        |> String.split("\n")
        |> Enum.map(&"      #{&1}")

      ["    command: |" | body]
    else
      ["    command: #{q(command)}"]
    end
  end

  defp step_env_lines(env) when map_size(env) == 0, do: []

  defp step_env_lines(env) do
    lines = env |> Map.keys() |> Enum.sort() |> Enum.map(&"      #{&1}: #{q(env[&1])}")
    ["    environment:" | lines]
  end

  # A JSON string literal, which is a valid YAML scalar by construction.
  defp q(s) when is_binary(s) do
    inner =
      s
      |> String.to_charlist()
      |> Enum.map_join(fn
        ?" -> "\\\""
        ?\\ -> "\\\\"
        ?\n -> "\\n"
        ?\r -> "\\r"
        ?\t -> "\\t"
        c when c < 0x20 -> "\\u" <> String.pad_leading(Integer.to_string(c, 16), 4, "0")
        c -> <<c::utf8>>
      end)

    "\"" <> inner <> "\""
  end

  @doc "Write into `<repo>/.tangled/workflows/` and return the path."
  @spec save(t(), Path.t()) :: {:ok, Path.t()} | {:error, term()}
  def save(%__MODULE__{} = wf, repo) do
    dir = Path.join([repo, ".tangled", "workflows"])

    with :ok <- File.mkdir_p(dir) do
      path = Path.join(dir, file_name(wf))

      content =
        "# Generated by the bsdkrun SDK — edit the code that save()d it instead.\n" <> yaml(wf)

      with :ok <- File.write(path, content), do: {:ok, path}
    end
  end

  @doc """
  Execute the workflow in a microVM, streaming output.

  The YAML never touches the repository — it goes to a temp file and
  `bsdkrun ci run -f`. Returns `:ok`, or `{:error, exit_code}` when a step
  fails.
  """
  @spec run(t(), keyword()) :: :ok | {:error, term()}
  def run(%__MODULE__{} = wf, opts \\ []) do
    tmp = Path.join(System.tmp_dir!(), "bsdkrun-ci-#{System.unique_integer([:positive])}")

    with :ok <- File.mkdir_p(tmp) do
      file = Path.join(tmp, file_name(wf))

      try do
        with :ok <- File.write(file, yaml(wf)) do
          args =
            ["ci", "run", "-f", file] ++
              case Keyword.get(opts, :dir) do
                nil -> []
                dir -> ["-w", dir]
              end

          bin = Bsdkrun.Binary.resolve!()
          {_out, code} = System.cmd(bin, args, into: IO.stream(:stdio, :line))
          if code == 0, do: :ok, else: {:error, code}
        end
      after
        File.rm_rf(tmp)
      end
    end
  end
end
